### PR TITLE
[Task 2] Configuring endpoints to manage APIs for several functionalities of the app in development. Added Unit tests and api doc using open api version 1.6.15.

### PR DESCRIPTION
**Pull request by minokabaivanov**
Pull request for review

### CODE REVIEW

File: ezGroceriesShoppingList/src/main/java/com/ezgroceries/shoppinglist/entity/Cocktail.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
+package com.ezgroceries.shoppinglist.entity;
+
+
+public class Cocktail {
```
Comment (DamianHristov) [inline ezGroceriesShoppingList/src/main/java/com/ezgroceries/shoppinglist/entity/Cocktail.java:16]: Add lombok and remove all getter setters and constructors.

File: ezGroceriesShoppingList/src/main/java/com/ezgroceries/shoppinglist/entity/Cocktail.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        this.ingredients = ingredients;
+    }
+
+    private String cocktailId;
+    private String name;
+
+    private String glass;
+    private String instructions;
+    private String image;
+    private String[] ingredients;// TODO convert to list
```
Comment (DamianHristov) [inline ezGroceriesShoppingList/src/main/java/com/ezgroceries/shoppinglist/entity/Cocktail.java]: As mentioned in the comment convert to list

File: ezGroceriesShoppingList/src/main/java/com/ezgroceries/shoppinglist/entity/ShoppingList.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+    private String shoppingListId;
+    private String name;
+
+    public ShoppingList(String shoppingListId, String name) {
+        this.shoppingListId = shoppingListId;
+        this.name = name;
+    }
+
+    private String[] ingredients;
```
Comment (DamianHristov) [inline ezGroceriesShoppingList/src/main/java/com/ezgroceries/shoppinglist/entity/ShoppingList.java]: For future references, if you join a team where there is no lombok and you must provide getters setter etc.. you should follow a structure
 -> first part of the class will be all the fields you provide 
 -> second part should be constructors 
 -> third part will be the getter setters and other functions related to your class

Comment (DamianHristov) [inline ezGroceriesShoppingList/src/main/java/com/ezgroceries/shoppinglist/entity/ShoppingList.java]: Convert to List

File: ezGroceriesShoppingList/src/main/java/com/ezgroceries/shoppinglist/entity/ShoppingList.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+    public ShoppingList(String shoppingListId, String name, String[] ingredients) {
+        this.shoppingListId = shoppingListId;
+        this.name = name;
+        this.ingredients = ingredients;
+    }
+
+    public ShoppingList(String name){
+        this.name=name;
```
Comment (DamianHristov) [inline ezGroceriesShoppingList/src/main/java/com/ezgroceries/shoppinglist/entity/ShoppingList.java]: space between '='

File: ezGroceriesShoppingList/src/main/java/com/ezgroceries/shoppinglist/service/CocktailService.java [hunk 5]
```diff
@@ -0,0 +1,12 @@
+package com.ezgroceries.shoppinglist.service;
+
+import com.ezgroceries.shoppinglist.entity.Cocktail;
+
+import java.util.List;
+
+public interface CocktailService {
+
+    List<Cocktail> listCocktails(String search);
```
Comment (DamianHristov) [inline ezGroceriesShoppingList/src/main/java/com/ezgroceries/shoppinglist/service/CocktailService.java]: Following naming convention either use the 'list' word inside the name of the function to indicate that it is not a single record OR use the plural form of words e.x.(cocktailS). But don't use both at the same time  
Imo  a better naming for this function will be: 
   -> getCoctails (I prefer this one) 
   ->getCoctailList

File: ezGroceriesShoppingList/src/main/java/com/ezgroceries/shoppinglist/util/TestData.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+                "https://www.thecocktaildb.com/images/media/drink/qtvvyq1439905913.jpg",
+                new String[]{"Tequila",
+                        "Blue Curacao",
+                        "Lime juice",
+                        "Salt"}
+        );
+        return List.of(firstCocktail, secondCocktail);
+    }
+
+    public static ShoppingList addCocktailToShoppingList() {
```
Comment (DamianHristov) [inline ezGroceriesShoppingList/src/main/java/com/ezgroceries/shoppinglist/util/TestData.java]: Why is this static method called add coctail to shipping list but actually returns a shopping list? 
Rename this method please

File: ezGroceriesShoppingList/src/main/java/com/ezgroceries/shoppinglist/util/TestData.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+        return new ShoppingList(
+                "Test1",
+                "Blue Margerita",
+                new String[]{"Tequila",
+                        "Blue Curacao",
+                        "Lime juice",
+                        "Salt"});
+    }
+
+    public static List<ShoppingList> allShoppingLists() {
```
Comment (DamianHristov) [inline ezGroceriesShoppingList/src/main/java/com/ezgroceries/shoppinglist/util/TestData.java]: add a get to function name

File: ezGroceriesShoppingList/src/main/java/com/ezgroceries/shoppinglist/web/WebController.java [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+    private static final Logger log = LoggerFactory.getLogger(WebController.class);
+
+
+    @Autowired
+    private CocktailService cocktailService;
+    @Autowired
+    private ShoppingListService shoppingListService;
+
+    @GetMapping("/cocktails")
+    @ResponseBody
```
Comment (DamianHristov) [inline ezGroceriesShoppingList/src/main/java/com/ezgroceries/shoppinglist/web/WebController.java]: remove this anotation

File: ezGroceriesShoppingList/src/main/java/com/ezgroceries/shoppinglist/web/WebController.java [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+    public List<Cocktail> getCocktails(@RequestParam String search) {
+        List<Cocktail> cocktails = cocktailService.listCocktails(search);
+        log.info("List of cocktails compiled");
+        return cocktails;
+
+    }
+
+    @PostMapping("/shopping-list")
+    public ResponseEntity<ShoppingList> createShoppingList(@RequestBody ShoppingList shoppingList) {
```
Comment (DamianHristov) [inline ezGroceriesShoppingList/src/main/java/com/ezgroceries/shoppinglist/web/WebController.java]: Understand the @valid anotation and inside shopping list class add anotation like @NotNull etc..

File: ezGroceriesShoppingList/src/main/java/com/ezgroceries/shoppinglist/web/WebController.java [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+
+    @GetMapping("/shopping-lists/{shoppingListId}")
+    @ResponseBody
+    public ShoppingList findShoppingListById(@PathVariable("shoppingListId") String shoppingListId) {
+        ShoppingList shoppingList = shoppingListService.findShoppingListById(shoppingListId);
+        return shoppingList;
+    }
+
+    @PostMapping("/shopping-lists/{shoppingListId}/cocktails")
+    public ResponseEntity<ShoppingList> addCocktailToShoppingList(@RequestBody Cocktail cocktail, @PathVariable("shoppingListId") String shoppingListId) {
```
Comment (DamianHristov) [inline ezGroceriesShoppingList/src/main/java/com/ezgroceries/shoppinglist/web/WebController.java]: Add @Valid and restriction for not null elements inside coctail class

File: ezGroceriesShoppingList/src/main/java/com/ezgroceries/shoppinglist/web/WebController.java [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+        String cocktailId = cocktail.getCocktailId();
```
Comment (DamianHristov) [inline ezGroceriesShoppingList/src/main/java/com/ezgroceries/shoppinglist/web/WebController.java]: There is no need to extract this coctailId in another variable, remove this line please

File: ezGroceriesShoppingList/src/main/java/com/ezgroceries/shoppinglist/web/WebController.java [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+        ShoppingList shoppingList = shoppingListService.addCocktailToShoppingList(shoppingListId, cocktailId);
+        log.info("Successfully added a cocktail " + cocktailId + " to shopping list " + shoppingListId);
+        return buildResponseEntity(shoppingList);
+    }
+
+    @GetMapping("/shopping-lists")
+    @ResponseBody
```
Comment (DamianHristov) [inline ezGroceriesShoppingList/src/main/java/com/ezgroceries/shoppinglist/web/WebController.java]: remove anotation

File: ezGroceriesShoppingList/src/main/java/com/ezgroceries/shoppinglist/web/WebController.java [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+    public List<ShoppingList> getAllShoppingLists() {
+        List<ShoppingList> shoppingLists = shoppingListService.findAllShoppingLists();
+        return shoppingLists;
+
```
Comment (DamianHristov) [inline ezGroceriesShoppingList/src/main/java/com/ezgroceries/shoppinglist/web/WebController.java]: Remove empty line

File: ezGroceriesShoppingList/src/test/java/com/ezgroceries/shoppinglist/EzGroceriesShoppingListApplicationTests.java [hunk 14]
```diff
@@ -1,11 +1,25 @@
...
 @SpringBootTest
 class EzGroceriesShoppingListApplicationTests {
 
+    private MockMvc mockMvc;
+    @Autowired
+    private WebApplicationContext context;
+
+    @Before
+    private void setUp() {
+        mockMvc = MockMvcBuilders.webAppContextSetup(context).build();
```
Comment (DamianHristov) [inline ezGroceriesShoppingList/src/test/java/com/ezgroceries/shoppinglist/EzGroceriesShoppingListApplicationTests.java]: Do you really need this?

File: ezGroceriesShoppingList/src/test/java/com/ezgroceries/shoppinglist/EzGroceriesShoppingListApplicationTests.java [hunk 15]
```diff
@@ -1,11 +1,25 @@
...
+    }
+
     @Test
     void contextLoads() {
```
Comment (DamianHristov) [inline ezGroceriesShoppingList/src/test/java/com/ezgroceries/shoppinglist/EzGroceriesShoppingListApplicationTests.java]: Why is this here

File: ezGroceriesShoppingList/src/test/java/com/ezgroceries/shoppinglist/WebControllerTests.java [hunk 16]
```diff
@@ -0,0 +1,129 @@
...
+public class WebControllerTests {
+
+
+    @MockBean
+    CocktailServiceImpl cocktailService;
+
+    @MockBean
+    ShoppingListService shoppingListService;
+
+    @SuppressWarnings("SpringJavaInjectionPointsAutowiringInspection")
```
Comment (DamianHristov) [inline ezGroceriesShoppingList/src/test/java/com/ezgroceries/shoppinglist/WebControllerTests.java:44]: Interesting

File: ezGroceriesShoppingList/src/test/java/com/ezgroceries/shoppinglist/WebControllerTests.java [hunk 17]
```diff
@@ -0,0 +1,129 @@
...
+
+        mockMvc.perform(get("/cocktails?search=test"))
+                .andExpect(status().isOk())
+                .andExpect(jsonPath("$", hasSize(2)))
+                .andExpect(jsonPath("$[0].name").value("Margerita"));
+
+        Mockito.verify(cocktailService, times(1)).listCocktails("test");
+    }
+
+    //Ignore the method below. It doesn't do anything. I created it before I realized that there is no direct way to test this method by url
```
Comment (DamianHristov) [inline ezGroceriesShoppingList/src/test/java/com/ezgroceries/shoppinglist/WebControllerTests.java:62]: Why is that

Comment (minokabaivanov) [reply ezGroceriesShoppingList/src/test/java/com/ezgroceries/shoppinglist/WebControllerTests.java:62]: Because this method can not be accessed by an endpoint. It is only used by another service method accessed by the endpoint to add a cocktail to a shopping list and the second one is already tested.

File: ezGroceriesShoppingList/src/test/java/com/ezgroceries/shoppinglist/WebControllerTests.java (hunk 16 above)
Comment (minokabaivanov) [reply ezGroceriesShoppingList/src/test/java/com/ezgroceries/shoppinglist/WebControllerTests.java:44]: It gave me an annoying warning that no beans are found. I researched it and learned that this is a problem for certain versions of Intellij and this is the solution.

File: ezGroceriesShoppingList/src/main/java/com/ezgroceries/shoppinglist/entity/Cocktail.java [hunk 18]
```diff
@@ -0,0 +1,26 @@
+package com.ezgroceries.shoppinglist.entity;
+
+
+import lombok.*;
```
Comment (minokabaivanov) [inline ezGroceriesShoppingList/src/main/java/com/ezgroceries/shoppinglist/entity/Cocktail.java:4]: Add option to intellij to block the wildcard import

File: ezGroceriesShoppingList/src/main/java/com/ezgroceries/shoppinglist/service/CocktailServiceImpl.java [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+import org.springframework.stereotype.Service;
+
+import java.util.ArrayList;
+import java.util.List;
+
+@Service
+public class CocktailServiceImpl implements CocktailService {
+
+    @Override
+    public List<Cocktail> getCocktails(String search) {
```
Comment (DamianHristov) [inline ezGroceriesShoppingList/src/main/java/com/ezgroceries/shoppinglist/service/CocktailServiceImpl.java:13]: Consistency is important, based on the naming convention you have chosen this method should be called findCocktails not getCocktails

File: ezGroceriesShoppingList/src/main/java/com/ezgroceries/shoppinglist/service/CocktailServiceImpl.java [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+                        add("Lime juice");
+                        add("Salt");
+                    }});
+        Cocktail secondCocktail = new Cocktail(
+                "d615ec78-fe93-467b-8d26-5d26d8eab073",
+                "Blue Margerita",
+                "Cocktail glass",
+                "Rub rim of cocktail glass with lime juice. Dip rim in coarse salt..",
+                "https://www.thecocktaildb.com/images/media/drink/qtvvyq1439905913.jpg",
+                new ArrayList() {
```
Comment (DamianHristov) [inline ezGroceriesShoppingList/src/main/java/com/ezgroceries/shoppinglist/service/CocktailServiceImpl.java:33]: One easier way to add objects to a list is the Method List.of(obj1, obj2, obj3). But that's fine also

File: ezGroceriesShoppingList/src/main/java/com/ezgroceries/shoppinglist/web/WebController.java [hunk 21]
```diff
@@ -0,0 +1,68 @@
...
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.*;
+
+import javax.validation.Valid;
+import java.util.List;
+
+@RestController
+@Slf4j
+public class WebController {
```
Comment (DamianHristov) [inline ezGroceriesShoppingList/src/main/java/com/ezgroceries/shoppinglist/web/WebController.java:21]: It's best to apply the separation of control principle. In this case we would have 2 controllers, one for cocktails, one for shopping carts. For now this is okay and we will leave it, but for the next PR, you should separate them

File: ezGroceriesShoppingList/src/test/java/com/ezgroceries/shoppinglist/EzGroceriesShoppingListApplicationTests.java [hunk 22]
```diff
@@ -1,13 +1,13 @@
...
+import org.junit.Before;
 import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.test.web.servlet.MockMvc;
+import org.springframework.test.web.servlet.setup.MockMvcBuilders;
+import org.springframework.web.context.WebApplicationContext;
 
 @SpringBootTest
 class EzGroceriesShoppingListApplicationTests {
```
Comment (DamianHristov) [inline ezGroceriesShoppingList/src/test/java/com/ezgroceries/shoppinglist/EzGroceriesShoppingListApplicationTests.java:12]: 404 Test not found

File: ezGroceriesShoppingList/src/test/java/com/ezgroceries/shoppinglist/WebControllerTests.java [hunk 23]
```diff
@@ -0,0 +1,134 @@
...
+                new ArrayList() {
+                    {
+                        add("Tequila");
+                        add("Blue Curacao");
+                        add("Lime juice");
+                        add("Salt");
+                    }});
+
+        when(cocktailService.findCocktailById(testId)).thenReturn(testCocktail);
+        Mockito.verify(cocktailService, times(0)).findCocktailById(testId);
```
Comment (DamianHristov) [inline ezGroceriesShoppingList/src/test/java/com/ezgroceries/shoppinglist/WebControllerTests.java:80]: you can use the never() method instead of times(0), but ignore this for now
